src/main/java/eu/europeana/keycloak/apikey/ApikeySynchronizer.java

```diff
@@ -0,0 +1,218 @@
+package eu.europeana.keycloak.apikey;
+
+import org.apache.http.Header;
+import org.apache.http.HttpStatus;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpDelete;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.client.methods.HttpRequestBase;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.message.BasicHeader;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.PreDestroy;
+import java.io.IOException;
+import java.util.Base64;
+
+/**
+ * Class used for synchronizing clients state with Apikey service. URL to Apikey service should be specified in the keycloak.json
+ * configuration file. So should be the credentials for the manager client authorized to perform synchronize calls.
+ *
+ */
+class ApikeySynchronizer {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ApikeySynchronizer.class);
+
+    // URL to Apikey service
+    private String apikeyServiceURL = null;
+
+    // Authorization header with manager client credentials used for each synchronize request
+    private Header authorizationHeader = null;
+
+    // Http client to perform calls to Apikey service
+    private CloseableHttpClient httpClient;
+
+    ApikeySynchronizer() {
+        httpClient = HttpClients.createDefault();
+    }
+
+    /**
+     * Initializes this object with data retrieved from configuration file. Client id and its secret
+     * are not stored but the authorization header is prepared based on them.
+     *
+     * @param apikeyServiceURL URL to Apikey service
+     * @param clientId client id of the manager client
+     * @param clientSecret secret corresponding to the client id
+     */
+    void init(String apikeyServiceURL, String clientId, String clientSecret) {
+        if (this.apikeyServiceURL == null) {
+            this.apikeyServiceURL = apikeyServiceURL;
+        }
+        if (this.authorizationHeader == null) {
+            this.authorizationHeader = prepareAuthorizationHeader(clientId, clientSecret);
+        }
+    }
+
+    /**
+     * In order to avoid storing credentials the authorization header is prepared and encoded. It is used for
+     * each call requiring basic client credentials authorization.
+     *
+     * @param clientId manager client id
+     * @param clientSecret secret corresponding to the clien
+     * @return Basic authorization header
+     */
+    private Header prepareAuthorizationHeader(String clientId, String clientSecret) {
+        return new BasicHeader("Authorization",
+                "Basic " + Base64.getEncoder().encodeToString((clientId + ":" + clientSecret).getBytes()));
```

Comment:
I am not very comfortable with using basic authentication because it's very insecure. Problem is that I'm not sure of a good alternative. I realize we only communicate between keycloak and apikeyservice so risk of sniffers is limited, but I do feel that if we are to use basic authentication, then it should be over SSL connections only. Could we add a check for this somehow?